monai/inferers/inferer.py

```diff
@@ -752,3 +771,1281 @@ def network_wrapper(
             return out
 
         return tuple(out_i.unsqueeze(dim=self.spatial_dim + 2) for out_i in out)
+
+
+class DiffusionInferer(Inferer):
+    """
+    DiffusionInferer takes a trained diffusion model and a scheduler and can be used to perform a signal forward pass
+    for a training iteration, and sample from the model.
```

Comment:
It may be confused here.
Looks like the inferer class added in this PR can be used for training as well. Do we really need these classes to inherit `Inferer`, or should we re-add a trainer for diffusion, or more detailed in the docstring?
https://github.com/Project-MONAI/GenerativeModels/blob/bdca1bd1ea8bda4960156285f805082fff7e8686/tutorials/generative/2d_controlnet/2d_controlnet.py#L246

cc @ericspod @Nic-Ma

Reply:
The aim was to simplify both the trainer and the evaluation of DDPMs. Doing sampling from a DDPM is more complicated than with other models, as it involves a combination of the model, a noise scheduler, and these have to be applied many times to produce outputs. So the initial motivation was to provide a [sampling method](https://github.com/Project-MONAI/MONAI/blob/a1e1bdad9f5b0e08fc48eb6f565ddc2a8a87142f/monai/inferers/inferer.py#L833) to make it easier to sample during validation/testing. 

We then added the `__call__` method to take care of some of the steps done during training the models (adding noise to the inputs and then evaluating the model). This becomes particularly helpful when training Latent DIffusion models, as the class also takes care of transforming to and from latent space for you, significantly simplifying your training loop.

So yes, these classes can be used during both training and inference, and break the mould of MONAI's `Infererer` class, in which `__call__` is used for inference, not trianing. The structure is also quite specific to the evaluation of diffusion models. Perhaps it is better they don't inherit from the `Inferer` class? They could either inherit from `nn.Module`  or we could create a new base class for generative model training/inference? I also agree the documentation needs to be changed to make their function clearer.

Reply:
I suggest not inheriting from the `Inferer` class and creating a new base class. What do you think? cc @marksgraham @ericspod @Nic-Ma 
But, of course, we could merge this one first.

Reply:
sure i can do a new base class. I've added it to the refactor issue. Would be great if we could merge this if everything else is OK - getting close to having everything moved over so soon we can start on the refactor!